### PR TITLE
[HDRP] density volumes improvements

### PR DESCRIPTION
**Pull request by alelievr**
### Purpose of this PR

⚠️ Waiting for Volumetric performance tests to land (ETA: ~2 weeks)

- Allows the Density Volume component to take a RenderTexture / Custom Render Texture as volume mask
- Add colored volume masks
- Higher-resolution volume masks (up to 256 cube configured in the HDRP settings)
- Added a falloff mode for density volume blend distance (linear or exponential)

Also, improve the 3D Texture atlas to support different 3D texture resolutions and RGBA 3D textures and real-time GPU copy for 3D render textures.

New volumetric settings in the HDRP asset:
image

Falloff mode:
image
Note that the falloff mode doesn't affect the Fade Distance interpolation (which is always linear).

Also added a debug mode to visualize the density volume 3D atlas:
image

With selection support:
image

---
### Testing status

Modified the fog volume test with 3D textures of different resolutions and colors:
image

https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/HDRP%252Fdensity-volumes-improvements/.yamato%252Fall-hdrp.yml%2523PR_HDRP_trunk/4610582/job/full-pipeline

### Note:
If you want to author texture 3D in unity I recommend this tool: https://github.com/alelievr/Mixture.
You can ask me for installation :)

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/DensityVolumeManager.cs
Comment (Wadarass) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/DensityVolumeManager.cs:65]: I might have missed it but does this mean the atlas is updated every frame now?  How do we detect if something has changed if not?

File: com.unity.render-pipelines.high-definition/Documentation~/Density-Volume.md
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Density-Volume.md:36]: There are two settings related to density volumes in the HDRP Asset settings: **Max Density Volume Size** and **Max Density Volume On Screen**. You can find both of these settings in the HDRP Asset's **Lighting** section.

File: com.unity.render-pipelines.high-definition/Documentation~/Density-Volume.md
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Density-Volume.md:40]: The **Max Density Volume Size** controls the maximum resolution you can use for the **Density Mask Texture**. Changing this setting can have a large impact on memory usage. For information on how much memory HDRP allocates for the Density Volume system, see the info box below this setting in the HDRP Asset.

File: com.unity.render-pipelines.high-definition/Documentation~/Density-Volume.md
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Density-Volume.md:42]: The **Max Density Volume On Screen** controls how many Density Volumes can appear on-screen at once. This setting also has an impact on memory.

File: com.unity.render-pipelines.high-definition/Documentation~/whats-new-11.md
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/whats-new-11.md:25]: Density Volumes masks now support using 3D RenderTextures as masks. 3D mask textures now also use all four RGBA channel which allows volumetric fog to have different colors and density based on the 3D Texture.

File: com.unity.render-pipelines.high-definition/Documentation~/whats-new-11.md
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/whats-new-11.md:27]: The size limit of 32x32x32 for the mask textures has also been replaced by a setting in the HDRP asset called "Max Density Volume Resolution", under the **Lighting > Volumetrics** section. The upper limit for mask textures is now 256x256x256, an info box below the field tells you how much memory is allocated to store these textures.

File: com.unity.render-pipelines.high-definition/Documentation~/whats-new-11.md
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/whats-new-11.md:29]: Finally, there is a new field to change the falloff HDRP applies when it blends the volume using the **Blend Distance** property. You can choose either **Linear** which is the default and previous technique, or **Exponential** which is more realistic.

**Review by JordanL8 — CHANGES_REQUESTED** [review]
Added some suggestions

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/DensityVolumeManager.cs
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/DensityVolumeManager.cs:65]: This only happens when the density volume texture is registered so no. Otherwise the it's the atlas that internally checks if the textures needs to be updated with a hash:
https://github.com/Unity-Technologies/Graphics/pull/2751/files#diff-9452a7b53ae106d6e8115c2ce8966e7dec933f97ce8d7241a18130cc5d9d4af7R373

**Comment by alelievr** [comment]
Pushed the doc fixes and added a bit I forgot in the what's new

**Review by JordanL8 — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/DensityVolume.cs
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/DensityVolume.cs:109]: Remove magic number

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumeVoxelization.compute
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumeVoxelization.compute:76]: Exponential is e^x, this is just a power (gamma) function.
Also: remove magic number.

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumeVoxelization.compute
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumeVoxelization.compute:87]: "Resolution" is ambiguous, use "size" or "dimensions".

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumeVoxelization.compute
Comment (EvgeniiG) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumeVoxelization.compute:109]: This looks expensive and could be optimized.

**Review by EvgeniiG — APPROVED** [review]
LGTM (after a very brief scan). Profile your changes.

**Comment by alelievr** [comment]
There will be a volumetric lighting performance test in the automated performance test project soon, we can wait for it to be merged and then see the perf impact with the dashboards

**Review by TomasKiniulis — COMMENTED** [review]
Looks good! I was hitting only one issue though:
Setting Max Density Volume Size to Resolution 256 throws the following:
`D3D11: Failed to create render texture resource (3D, 2048x8192x256 mips=14 dxfmt=27, hr=0x30)`, checking with Antoine a user could be informed with a warning notifying that gpu memory limit is reached

 Went through settings, debug modes, exponential/linear settings. No other issues found.

File: com.unity.render-pipelines.high-definition/Documentation~/Density-Volume.md
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Documentation~/Density-Volume.md:46]: guess it need to be rewrite as it is not grayscale anymore, we handle RGB

File: com.unity.render-pipelines.high-definition/Documentation~/whats-new-11.md
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Documentation~/whats-new-11.md:27]: Guess we should say in the doc that density resoltion and froxel resolution are two different concept, so putting a 256 texture don't necessary improve resolution if the froxel don't change. All depends on the relative size between volume and froxel.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipelineResources/HDRenderPipelineResources.asset
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipelineResources/HDRenderPipelineResources.asset:224]: We have lost VTFeedbackDownsample, this need to be re-introduce. Note: this is a problem I have solve on master, so if you have latest master and regenerate resources it will works.

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipelineResources/HDRenderPipelineResources.asset:224]: so you need to regenerate it in the PR

File: com.unity.render-pipelines.high-definition/Documentation~/whats-new-11.md
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Documentation~/whats-new-11.md:61]: Yes, indeed. I'll add it

**Comment by alelievr** [comment]
@TomasKiniulis I fixed the texture failed to create issue

MaxSize

**Comment by TomasKiniulis** [comment]
> @TomasKiniulis I fixed the texture failed to create issue
> 
> MaxSize

Thanks! Seems like a nice way to avoid the issue

**Review by TomasKiniulis — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumeVoxelization.compute
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/VolumeVoxelization.compute:78]: PositivePow

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Utility/Texture3DAtlas.compute
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Utility/Texture3DAtlas.compute:3]: caution

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/DebugDensityVolumeAtlas.shader
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/DebugDensityVolumeAtlas.shader:12]: ici

**Pull request merged**
